src/semaphore.jl

```diff
@@ -0,0 +1,29 @@
+using ConcurrentSim
+using ResumableFunctions
+import Base: unlock, lock
+
+"""Multiple processes can wait on this semaphore for a permission to run given by another process"""
+struct AsymmetricSemaphore
+    nbwaiters::Ref{Int}
+    lock::Resource
+end
+AsymmetricSemaphore(sim) = AsymmetricSemaphore(Ref(0), Resource(sim,1,level=1)) # start locked
+
+function Base.lock(s::AsymmetricSemaphore)
+    return @process _lock(s.lock.env, s)
+end
+
+@resumable function _lock(sim, s::AsymmetricSemaphore)
+    s.nbwaiters[] += 1
+    @yield lock(s.lock)
+    s.nbwaiters[] -= 1
+    if s.nbwaiters[] > 0
+        unlock(s.lock)
+    end
+end
+
+function unlock(s::AsymmetricSemaphore)
+    if s.nbwaiters[] > 0
+        unlock(s.lock)
+    end
+end
```

Comment:
The rest of the library does not automatically know about the existence of this file. In needs to be included from the main file `src/QuantumSavory.jl`. Given that it is a dependency for the content of `states_registers.jl` it should probably be included before it. E.g. here: https://github.com/QuantumSavory/QuantumSavory.jl/blob/46cb2f30b0e16167f12ca63082f11c6a8c0f3303/src/QuantumSavory.jl#L82